tools/base/utils.py

```diff
@@ -0,0 +1,28 @@
+#
+# Provides shared utils used by other python modules
+#
+
+import os
+import tempfile
+from configparser import ConfigParser
+from contextlib import contextmanager
+from typing import Iterator
+
+
+# this is testing specific - consider moving to tools.testing.utils
+@contextmanager
+def custom_coverage_data(data_file: str) -> Iterator[str]:
```

Comment:
I don't understand why we need to have this customization? Why isn't coverage push button?

Reply:
afaict coverage doesnt take the `data_file` arg as an arg - only from the config - so i had to do this config mangling to make it work

Reply:
I don't quite get how the data file figures into using coverage..

Reply:
when you run coverage it creates a data file - its the same format for all langs - so you can create coverage reports from multiple langs

when you run coverage (or invoke it  with some test runner like pytest) it can either create or append to an existing file

running lots of pytests with bazel means it has to be passed between the test runs

at the end this implementation invokes coverage directly to generate html from the data file

Reply:
ive renamed this to `coverage_with_data_file` which hopefully is a bit clearer